### PR TITLE
Better uphill slope RNG fix

### PR DESCRIPTION
**Pull request by jason-e**
Hello! I made [RNGFix](https://forums.alliedmods.net/showthread.php?t=310825) which aims to make psuedo-random physics behavior behave consistently. Momentum already contains fixes for some of these things, but not all (yet!), and this PR focuses on the uphill slopes issue.

An explanation of the issue for those unaware:

At the end of a tick of physics simulation, if the player is within 2 units of a shallow-enough surface and is not moving up too fast, they are immediately considered "on the ground", which sets their z-velocity to zero. When this happens, it means the player never actually collides with the ground and ends up landing (and jumping) as if on totally flat ground. Whether or not this happens *essentially* comes down to chance -- it just depends on how the ticks happen to line up with the ground.

So the player will either collide with the ground first, or "land" first. When hitting an incline going "downhill", the player would rather collide first, because doing so results in their z-velocity being turned into some horizontal velocity, which is often the purpose of such inclines (Momentum already makes sure this happens). While it does make sense from a purely physical perspective for the player to also always collide first when going *uphill* into an incline, landing first is actually more beneficial to the player as it helps them retain more speed.

The fact is that ever since game modes that support bunny hopping like bhop and surf have been around, players have played on an engine that gave them that chance -- often not very small -- to "land" first when going up slopes and keep more speed. It has been used in many record runs, and has implications on many scenarios including launching off slopes, but also just bhopping on shallow inclines or bumpy ground. 

I think anyone would agree that random chance has no place in a skill-focused game, and if we break down the choice as either giving the player "good RNG" every time, or giving them "bad RNG" every time, to me the clear choice is "good RNG". A more long-winded version of my philosophy on this can be found [here](https://pastebin.com/raw/WTYy7077).

---

This implementation of the fix is smarter and cleaner than the current public version of RNGFix. During the main part of air movement simulation (`TryPlayerMove()`), it checks if the player is about to collide with an incline that they *could have* landed on -- given the game's requirements on landing -- had they been luckier and things lined up differently. When it sees this, it stops the movement simulation part of the tick immediately. This allows other routines `CategorizePosition()`, and later `CheckJumpButton()` on the following tick, to be run before movement continues and the collision is able to occur. This method has been in-use in my newer, beta version of RNGFix on Horizon Servers for over a year now, so I'm confident in it's correctness, barring perhaps some obscure interaction with Momentum's ramp bug fix.

Finally, once the engine has been changed to always make the player land first when going uphill, another engine quirk reveals itself more clearly: Hitting an incline while *not* holding jump often gives the player more speed than if they had held jump. This is both unintuitive and something that could more clearly be called a bug, as the increase in speed essentially comes from nowhere due to how the `StepMove()` function acts (it wasn't really designed for bhop speeds). This PR also adds a ConVar, `sv_nojump_slope_boosts` which can be turned off to enable an experimental tweak to `StepMove()` that keeps the player from being given extra speed from nowhere. Because this boost effect is entirely consistent, I wouldn't say preventing it is *obviously* necessary, but I included the fix because I didn't want the approval of the main part of this PR from being held back by this secondary issue.

### Checklist
- [x] If there is a localization token change, I have updated the `momentum_english_ref.res` file with the changes, ran `tokenizer.py` to generate an up-to-date localization file, and have committed both the `.res` file changes and the new localization `.txt` file
- [x] If I introduced new h/cpp files, I have added them to the appropriate project's VPC file (`server_momentum.vpc` / `client_momentum.vpc` / etc)
- [x] My commits are relatively small and scoped to the best of my ability
- [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
- [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review

### CODE REVIEW

File: mp/src/game/shared/momentum/mom_gamemovement.cpp [hunk 1]
```diff
@@ -328,6 +331,119 @@ void CMomentumGameMovement::WalkMove()
...
+
+    // Copy this origin to up.
+    Vector vecUpPos;
+    VectorCopy(mv->GetAbsOrigin(), vecUpPos);
+
+    // decide which one went farther
+    float flDownDist =
+        (vecDownPos.x - vecPos.x) * (vecDownPos.x - vecPos.x) + (vecDownPos.y - vecPos.y) * (vecDownPos.y - vecPos.y);
+    float flUpDist =
+        (vecUpPos.x - vecPos.x) * (vecUpPos.x - vecPos.x) + (vecUpPos.y - vecPos.y) * (vecUpPos.y - vecPos.y);
```
Comment (Gocnak) [inline mp/src/game/shared/momentum/mom_gamemovement.cpp:424]:
```suggestion
    float flDownDist = vecDownPos.DistToSqr(vecPos);
    float flUpDist = vecUpPos.DistToSqr(vecPos);
```

File: mp/src/game/shared/momentum/mom_gamemovement.cpp [hunk 2]
```diff
@@ -37,6 +38,8 @@
 // remove this eventually
 ConVar sv_slope_fix("sv_slope_fix", "1");
 ConVar sv_ramp_fix("sv_ramp_fix", "1");
+ConVar sv_nojump_slope_boosts("sv_nojump_slope_boosts", "1", 0,
+    "Keep original engine behavior which makes going up slopes without holding jump usually result in extra speed.");
```
Comment (Gocnak) [inline mp/src/game/shared/momentum/mom_gamemovement.cpp:42]: As per Aux's request, you may remove this and treat the code that uses it as if it were hardcoded to 0.

File: mp/src/game/shared/momentum/mom_gamemovement.cpp [hunk 3]
```diff
@@ -1286,20 +1402,43 @@ void CMomentumGameMovement::CategorizePosition()
...
+                            SetGroundEntity(&pm);
+                        }
+                        else
+                        {
+                            // Pretend the player collided first and since that means they can't land,
+                            // don't set the ground entity
+
+                            // Should we also pre-apply the ClipVelocity result?
+                            // VectorCopy(vecNextVelocity, mv->m_vecVelocity);
+                        }
```
Comment (Gocnak) [inline mp/src/game/shared/momentum/mom_gamemovement.cpp:1436]: Empty block? You can remove this and leave the comment to explain why there's no else.

File: mp/src/game/shared/momentum/mom_gamemovement.cpp [hunk 4]
```diff
@@ -1887,6 +2026,44 @@ int CMomentumGameMovement::TryPlayerMove(Vector *pFirstDest, trace_t *pFirstTrac
             else
             {
                 TracePlayerBBox(mv->GetAbsOrigin(), end, PlayerSolidMask(), COLLISION_GROUP_PLAYER_MOVEMENT, pm);
+
+                // This fix might also need to be applied after the sv_ramp_fix TracePlayerBBox above...
```
Comment (Gocnak) [inline mp/src/game/shared/momentum/mom_gamemovement.cpp:2030]: As per @chrb22's review, this comment is unnecessary and can be removed

File: mp/src/game/shared/momentum/mom_gamemovement.cpp (hunk 1 above)
Comment (chrb22) [reply mp/src/game/shared/momentum/mom_gamemovement.cpp:424]: Should be 2D, so flDownDist = (vecDownPos - vecPos).Length2DSqr() and similar.

Comment (Gocnak) [reply mp/src/game/shared/momentum/mom_gamemovement.cpp:424]: Good catch! Surprised there isn't a `DistTo2D`...

Comment (jason-e) [reply mp/src/game/shared/momentum/mom_gamemovement.cpp:424]: The bulk of that StepMove override is the original code from CGameMovement, I only changed a couple things where you can see the convar checked. It wouldn't hurt to change it to use that briefer distance calculation you suggested, but technically that method also subtracts the z components unnecessarily.

File: mp/src/game/shared/momentum/mom_gamemovement.cpp (hunk 3 above)
Comment (jason-e) [reply mp/src/game/shared/momentum/mom_gamemovement.cpp:1436]: I'll do that now. Wasn't totally set on if that copy operation should actually be there or not but it should be fine leaving it out.

**Pull request merged**